llama-index-integrations/llms/llama-index-llms-keywordsai/llama_index/llms/keywordsai/base.py

```diff
@@ -0,0 +1,971 @@
+    acompletion_to_chat_decorator,
+    astream_chat_to_completion_decorator,
+    astream_completion_to_chat_decorator,
+    chat_to_completion_decorator,
+    completion_to_chat_decorator,
+    stream_chat_to_completion_decorator,
+    stream_completion_to_chat_decorator,
+)
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+    ChatResponseAsyncGen,
+    ChatResponseGen,
+    CompletionResponse,
+    CompletionResponseAsyncGen,
+    CompletionResponseGen,
+    LLMMetadata,
+    MessageRole,
+)
+from llama_index.core.bridge.pydantic import Field, PrivateAttr
+from llama_index.core.callbacks import CallbackManager
+from llama_index.core.constants import (
+    DEFAULT_TEMPERATURE,
+)
+from llama_index.core.llms.callbacks import (
+    llm_chat_callback,
+    llm_completion_callback,
+)
+from llama_index.core.llms.llm import ToolSelection
+from llama_index.core.types import BaseOutputParser, PydanticProgramMode, Model
+from .utils import (
+    O1_MODELS,
+    KeywordsAIToolCall,
+    create_retry_decorator,
+    from_openai_completion_logprobs,
+    from_openai_message,
+    from_openai_token_logprobs,
+    is_chat_model,
+    is_function_calling_model,
+    keywordsai_modelname_to_contextsize,
+    resolve_openai_credentials,
+    to_openai_message_dicts,
+    resolve_tool_choice,
+    update_tool_calls,
+)
+from llama_index.core.bridge.pydantic import (
+    BaseModel,
+)
+
+from openai import AsyncOpenAI, AzureOpenAI
+from openai import OpenAI as SyncOpenAI
+from openai.types.chat.chat_completion_chunk import (
+    ChatCompletionChunk,
+    ChoiceDelta,
+    ChoiceDeltaToolCall,
+)
+from llama_index.core.llms.utils import parse_partial_json
+
+import llama_index.core.instrumentation as instrument
+
+dispatcher = instrument.get_dispatcher(__name__)
+
+if TYPE_CHECKING:
+    from llama_index.core.tools.types import BaseTool
+
+DEFAULT_KEYWORDSAI_MODEL = "gpt-4o"
+
+
+def llm_retry_decorator(f: Callable[[Any], Any]) -> Callable[[Any], Any]:
+    @functools.wraps(f)
+    def wrapper(self, *args: Any, **kwargs: Any) -> Any:
+        max_retries = getattr(self, "max_retries", 0)
+        if max_retries <= 0:
+            return f(self, *args, **kwargs)
+
+        retry = create_retry_decorator(
+            max_retries=max_retries,
+            random_exponential=True,
+            stop_after_delay_seconds=60,
+            min_seconds=1,
+            max_seconds=20,
+        )
+        return retry(f)(self, *args, **kwargs)
+
+    return wrapper
+
+
+@runtime_checkable
+class Tokenizer(Protocol):
+    """Tokenizers support an encode function that returns a list of ints."""
+
+    def encode(self, text: str) -> List[int]:
+        ...
+
+
+def force_single_tool_call(response: ChatResponse) -> None:
+    tool_calls = response.message.additional_kwargs.get("tool_calls", [])
+    if len(tool_calls) > 1:
+        response.message.additional_kwargs["tool_calls"] = [tool_calls[0]]
+
+
+class KeywordsAI(OpenAI):
```

Comment:
High-level -- users should use `OpenAILike` to access openai-like APIs. This would probably avoid some of the checks/errors that you noted in the PR description

That being said, happy to have a specific integration for keywords ai. The implementation likely can be a lot smaller though, most methods are copies from the openai class right?